lncfg/db.go

```diff
@@ -79,54 +107,267 @@ func (db *DB) Init(ctx context.Context, dbPath string) error {
+	walletDBPath, towerServerDBPath string, towerClientEnabled,
+	towerServerEnabled bool) (*DatabaseBackends, error) {
 
-	var (
-		localDB, remoteDB kvdb.Backend
-		err               error
-	)
+	// We keep track of all the kvdb backends we actually open and return a
+	// reference to their close function so they can be cleaned up properly
+	// on error or shutdown.
+	closeFuncs := make(map[string]func() error)
+
+	// If we need to return early because of an error, we invoke any close
+	// function that has been initialized so far.
+	returnEarly := true
+	defer func() {
+		if !returnEarly {
+			return
+		}
+
+		for _, closeFunc := range closeFuncs {
+			_ = closeFunc()
+		}
+	}()
 
 	if db.Backend == EtcdBackend {
-		remoteDB, err = kvdb.Open(
-			kvdb.EtcdBackendName, ctx, db.Etcd,
+		// As long as the graph data, channel state and height hint
+		// cache are all still in the channel.db file in bolt, we
+		// replicate the same behavior here and use the same etcd
+		// backend for those three sub DBs. But we namespace it properly
+		// to make such a split even easier in the future. This will
+		// break lnd for users that ran on etcd with 0.13.x since that
+		// code used the root namespace. We assume that nobody used etcd
+		// for mainnet just yet since that feature was clearly marked as
+		// experimental in 0.13.x.
+		etcdBackend, err := kvdb.Open(
+			kvdb.EtcdBackendName, ctx,
+			db.Etcd.CloneWithSubNamespace(NSChannelDB),
+		)
+		if err != nil {
+			return nil, fmt.Errorf("error opening etcd DB: %v", err)
+		}
+		closeFuncs[NSChannelDB] = etcdBackend.Close
+
+		etcdMacaroonBackend, err := kvdb.Open(
+			kvdb.EtcdBackendName, ctx,
+			db.Etcd.CloneWithSubNamespace(NSMacaroonDB),
+		)
+		if err != nil {
+			return nil, fmt.Errorf("error opening etcd macaroon "+
+				"DB: %v", err)
+		}
+		closeFuncs[NSMacaroonDB] = etcdMacaroonBackend.Close
+
+		etcdDecayedLogBackend, err := kvdb.Open(
+			kvdb.EtcdBackendName, ctx,
+			db.Etcd.CloneWithSubNamespace(NSDecayedLogDB),
+		)
+		if err != nil {
+			return nil, fmt.Errorf("error opening etcd decayed "+
+				"log DB: %v", err)
+		}
+		closeFuncs[NSDecayedLogDB] = etcdDecayedLogBackend.Close
+
+		etcdTowerClientBackend, err := kvdb.Open(
+			kvdb.EtcdBackendName, ctx,
+			db.Etcd.CloneWithSubNamespace(NSTowerClientDB),
+		)
+		if err != nil {
+			return nil, fmt.Errorf("error opening etcd tower "+
+				"client DB: %v", err)
+		}
+		closeFuncs[NSTowerClientDB] = etcdTowerClientBackend.Close
+
+		etcdTowerServerBackend, err := kvdb.Open(
+			kvdb.EtcdBackendName, ctx,
+			db.Etcd.CloneWithSubNamespace(NSTowerServerDB),
 		)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("error opening etcd tower "+
+				"server DB: %v", err)
 		}
+		closeFuncs[NSTowerServerDB] = etcdTowerServerBackend.Close
+
+		etcdWalletBackend, err := kvdb.Open(
+			kvdb.EtcdBackendName, ctx,
+			db.Etcd.
+				CloneWithSubNamespace(NSWalletDB).
+				CloneWithSingleWriter(),
```

Comment:
I still feel pretty nervous about just flipping all of the databases besides this one to concurrent writing. Itests may pass, but in general those don't really test concurrent scenarios. How certain is it that those would catch concurrency issues?

Maybe it is worth considering to start out more conservative here. Things like macaroondb and the towerdbs - do those really need concurrent writing from the start?

As mentioned elsewhere, I'd prefer to not do any concurrent writing at all for the first etcd release.

Reply:
Discussed offline. We'll add more bug fixes and tests on top of this PR to make sure everything is concurrency safe. Also having multiple weeks of Travis runs with the lock removed should help us find at least some of those hidden bugs (that we want to fix anyway in the mid term).